### PR TITLE
Add spread operator for array initializers, function calls, and method calls

### PR DESCRIPTION
**Pull request by KevinEady**
TODO:
- [x] core changes
- [x] escript guide
- [x] more tests?

### CODE REVIEW

File: pol-core/bscript/compiler/ast/SpreadElement.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+void SpreadElement::accept( NodeVisitor& visitor )
+{
+  visitor.visit_spread_element( *this );
+}
+
+void SpreadElement::describe_to( std::string& w ) const
+{
+  fmt::format_to( std::back_inserter( w ), "string-value({})", "!" );
```
Comment (KevinEady) [inline pol-core/bscript/compiler/ast/SpreadElement.cpp:19]: c&p

File: pol-core/bscript/executor.cpp [hunk 2]
```diff
@@ -2991,6 +3007,40 @@ void Executor::ins_dictionary( const Instruction& /*ins*/ )
...
+  }
+
+  auto arr = static_cast<ObjArray*>( ref->impptr() );
+
+  if ( mode == 0 )
+  {
+    passert_always( !ValueStack.empty() );
+    auto& stack_arr_ref = ValueStack.back();
+    passert_always( stack_arr_ref->impptr()->isa( BObjectImp::OTArray ) );
+    auto stack_arr = static_cast<ObjArray*>( stack_arr_ref->impptr() );
```
Comment (KevinEady) [inline pol-core/bscript/executor.cpp:3035]: use new syntax

File: pol-core/bscript/compiler/codegen/InstructionGenerator.cpp [hunk 3]
```diff
@@ -101,11 +102,25 @@ void InstructionGenerator::visit_array_initializer( ArrayInitializer& node )
 {
   update_debug_location( node );
   emit.array_create();
+
+  // Spreaded elements go into the current array for array initializers, eg. in
+  // `{ ...spread }`.
```
Comment (KevinEady) [inline pol-core/bscript/compiler/codegen/InstructionGenerator.cpp:107]: change syntax in doc

File: pol-core/bscript/compiler/codegen/InstructionGenerator.cpp [hunk 4]
```diff
@@ -389,9 +404,13 @@ void InstructionGenerator::visit_function_call( FunctionCall& call )
   // A function call will have no link if it is an expression call, eg. `(foo)(1,2)`.
   if ( !call.function_link->function() )
   {
+    // Spreaded elements go into the value stack for expression-as-callee
+    // function calls, eg. in `(foo)(...spread)`.
```
Comment (KevinEady) [inline pol-core/bscript/compiler/codegen/InstructionGenerator.cpp:408]: change syntax in doc

File: pol-core/bscript/compiler/codegen/InstructionGenerator.cpp [hunk 5]
```diff
@@ -641,7 +675,11 @@ void InstructionGenerator::visit_member_assignment_by_operator( MemberAssignment
 
 void InstructionGenerator::visit_method_call( MethodCall& method_call )
 {
+  // Spreaded elements go into the value stack in method calls, in
+  // eg. `foo.bar(...spread)`
```
Comment (KevinEady) [inline pol-core/bscript/compiler/codegen/InstructionGenerator.cpp:683]: change syntax in doc

File: pol-core/bscript/compiler/codegen/InstructionGenerator.cpp [hunk 6]
```diff
@@ -710,6 +748,19 @@ void InstructionGenerator::visit_return_statement( ReturnStatement& ret )
...
 }
 
+void InstructionGenerator::visit_spread_element( SpreadElement& node )
+{
+  if ( spread_modes.empty() )
+  {
+    node.internal_error( "spread element outside of array initializer" );
+    return;
+  }
+  // auto mode = spread_modes.empty() ? SpreadMode::Array : spread_modes.top();
```
Comment (KevinEady) [inline pol-core/bscript/compiler/codegen/InstructionGenerator.cpp:765]: remove

File: pol-core/bscript/executor.cpp [hunk 7]
```diff
@@ -2991,6 +3007,40 @@ void Executor::ins_dictionary( const Instruction& /*ins*/ )
...
+        "Error in script '{}':\n"
+        "\tAttempt to use spread operator on non-array.",
+        prog_->name, PC );
+
+    return;
+  }
+
+  auto arr = static_cast<ObjArray*>( ref->impptr() );
+
+  if ( mode == 0 )
```
Comment (KevinEady) [inline pol-core/bscript/executor.cpp:3030]: document this more

File: pol-core/bscript/executor.cpp (hunk 2 above)
Comment (KevinEady) [reply pol-core/bscript/executor.cpp:3035]: does not exist in bspread implementation

File: pol-core/bscript/compiler/codegen/InstructionGenerator.cpp (hunk 3 above)
Comment (KevinEady) [reply pol-core/bscript/compiler/codegen/InstructionGenerator.cpp:107]: does not exist in bspread implementation

File: pol-core/bscript/compiler/codegen/InstructionGenerator.cpp (hunk 4 above)
Comment (KevinEady) [reply pol-core/bscript/compiler/codegen/InstructionGenerator.cpp:408]: does not exist in bspread implementation

File: pol-core/bscript/compiler/codegen/InstructionGenerator.cpp (hunk 5 above)
Comment (KevinEady) [reply pol-core/bscript/compiler/codegen/InstructionGenerator.cpp:683]: does not exist in bspread implementation

File: pol-core/bscript/executor.cpp (hunk 7 above)
Comment (KevinEady) [reply pol-core/bscript/executor.cpp:3030]: does not exist in bspread implementation

File: pol-core/bscript/executor.cpp [hunk 8]
```diff
@@ -306,9 +306,42 @@ int Executor::getParams( unsigned howMany )
       ValueStack.pop_back();
     }
   }
+  expandParams();
   return 0;
 }
 
+void Executor::expandParams()
+{
+  for ( auto i = static_cast<int>( fparams.size() ) - 1; i >= 0; --i )
```
Comment (turleypol) [inline pol-core/bscript/executor.cpp:315]: At least in theory fparams.size could be zero

File: pol-core/bscript/object.cpp [hunk 9]
```diff
@@ -2395,15 +2395,58 @@ bool BFunctionRef::variadic() const
...
 }
 
-BObjectImp* BFunctionRef::call_method_id( const int id, Executor& /*ex*/, bool /*forcebuiltin*/ )
+BObjectImp* BFunctionRef::call_method_id( const int id, Executor& ex, bool /*forcebuiltin*/ )
 {
   switch ( id )
   {
+    // This is only entered if `ins_call_method_id` did _not_ do the call jump.
   case MTH_CALL:
-    return nullptr;  // handled directly
```
Comment (turleypol) [inline pol-core/bscript/object.cpp:2403]: Isn't the comment still true? At least without knowing where it was, but it's not part of the diff

Comment (KevinEady) [reply pol-core/bscript/object.cpp:2403]: Yes, but I clarified it a bit with the above comment
https://github.com/polserver/polserver/blob/5f7dd8dde48a070acf0e48b667cf1e3bc143b5eb/pol-core/bscript/object.cpp#L2402

**Review by turleypol — APPROVED** [review]

**Pull request merged**